spark/src/test/scala/org/apache/spark/sql/delta/clustering/ClusteringTableFeatureSuite.scala

```diff
@@ -49,24 +60,60 @@ class ClusteringTableFeatureSuite extends SparkFunSuite with DeltaSQLCommandTest
     }
   }
 
-  test("alter table cluster by non-clustered tables is not allowed.") {
+  test("alter table cluster by partitioned tables is not allowed.") {
     withTable("tbl") {
-      sql("CREATE TABLE tbl(a INT, b STRING) USING DELTA")
+      sql("CREATE TABLE tbl(a INT, b STRING) USING DELTA PARTITIONED BY (a)")
       val e1 = intercept[DeltaAnalysisException] {
         sql("ALTER TABLE tbl CLUSTER BY (a)")
       }
       checkError(
         e1,
-        "DELTA_ALTER_TABLE_CLUSTER_BY_NOT_ALLOWED",
+        "DELTA_ALTER_TABLE_CLUSTER_BY_ON_PARTITIONED_TABLE_NOT_ALLOWED",
         parameters = Map.empty)
 
       val e2 = intercept[DeltaAnalysisException] {
         sql("ALTER TABLE tbl CLUSTER BY NONE")
       }
       checkError(
         e2,
-        "DELTA_ALTER_TABLE_CLUSTER_BY_NOT_ALLOWED",
+        "DELTA_ALTER_TABLE_CLUSTER_BY_ON_PARTITIONED_TABLE_NOT_ALLOWED",
         parameters = Map.empty)
     }
   }
+
+   test("alter table cluster by unpartitioned non-clustered tables is supported.") {
+    val table = "tbl"
+    withTable(table) {
+      sql(s"CREATE TABLE $table (a INT, b STRING) USING DELTA")
+      val clusterByLogs = Log4jUsageLogger.track {
+        sql(s"ALTER TABLE $table CLUSTER BY (a)")
+      }.filter { e =>
+        e.metric == MetricDefinitions.EVENT_TAHOE.name &&
+          e.tags.get("opType").contains("delta.ddl.alter.clusterBy")
+      }
+      assert(clusterByLogs.nonEmpty)
+      val clusterByLogJson = JsonUtils.fromJson[Map[String, Any]](clusterByLogs.head.blob)
+      assert(!clusterByLogJson("isClusterByNoneSkipped").asInstanceOf[Boolean])
+      val (_, finalSnapshot) = DeltaLog.forTableWithSnapshot(spark, TableIdentifier(table))
+      assert(ClusteredTableUtils.isSupported(finalSnapshot.protocol))
+
+      withSQLConf(SQLConf.MAX_RECORDS_PER_FILE.key -> "2") {
+        val df = (1 to 4).map(i => (i, i.toString)).toDF("a", "b")
+        withSQLConf(SQLConf.MAX_RECORDS_PER_FILE.key -> "1") {
+          df.write.format("delta").mode("append").saveAsTable(table)
+        }
+
+        // Optimize should cluster the data into two 2 files since MAX_RECORDS_PER_FILE is 2.
+        runOptimize(table) { metrics =>
+          assert(metrics.numFilesRemoved == 4)
+          assert(metrics.numFilesAdded == 2)
+        }
+
+        val deltaLog = DeltaLog.forTable(spark, TableIdentifier(table))
+        val files1 = deltaLog.update().allFiles.collect().toSet
+        assert(files1.size == 2)
+        assert(files1.forall(_.clusteringProvider.contains(ClusteredTableUtils.clusteringProvider)))
```

Comment:
Can we add a check the column `a` is sorted?

Reply:
We fallback to zorder for 1 column, which doesn't always guarantee sorted order within the same file. From my testing it can also show both [1, 2], [3, 4] and [2, 1], [3, 4] for the two files.

Reply:
Got it, how about checking the min-max interval are sorted?

Reply:
Sure, added the check